test/src/org/apache/jmeter/functions/TestTimeShiftingFunction.java

```diff
@@ -0,0 +1,161 @@
+import org.junit.Test;
+
+public class TestTimeShiftingFunction extends JMeterTestCase {
+        private Function function;
+
+        private SampleResult result;
+
+        private Collection<CompoundVariable> params;
+
+        private JMeterVariables vars;
+
+        private JMeterContext jmctx = null;
+
+        private String value;
+        
+        @Before
+        public void setUp() {
+            jmctx = JMeterContextService.getContext();
+            vars = new JMeterVariables();
+            jmctx.setVariables(vars);
+            jmctx.setPreviousResult(result);
+            params = new LinkedList<>();
+            result = new SampleResult();
+            function = new TimeShiftingFunction();
+        }
+
+        @Test
+        public void testDatePlusOneDay() throws Exception {
+            Collection<CompoundVariable> params = makeParams("yyyy-dd-MM","2017-01-01","1d","");
+            function.setParameters(params);
+            value = function.execute(result, null);
+            assertTrue(value.equalsIgnoreCase("2017-02-01"));
+        }
+        
+        @Test
+        public void testDatePlusOneDayInVariable() throws Exception {
+            Collection<CompoundVariable> params = makeParams("yyyy-dd-MM","2017-01-01","1d","VAR");
+            function.setParameters(params);
+            function.execute(result, null);
+            assertTrue(vars.get("VAR").equalsIgnoreCase("2017-02-01"));
+        }
+        
+        @Test
+        public void testDateLessOneDay() throws Exception {
+            Collection<CompoundVariable> params = makeParams("yyyy-dd-MM","2017-01-01","-1d","");
+            function.setParameters(params);
+            value = function.execute(result, null);
+            assertTrue(value.equalsIgnoreCase("2016-31-12"));
+        }
+        
+        @Test
+        public void testDatePlusOneHour() throws Exception {
+            Collection<CompoundVariable> params = makeParams("HH:mm:ss","14:00:00","1H","");
+            function.setParameters(params);
+            value = function.execute(result, null);
+            assertTrue(value.equalsIgnoreCase("15:00:00"));
+        }
+        
+        @Test
+        public void testDateLessOneMinute() throws Exception {
+            Collection<CompoundVariable> params = makeParams("HH:mm:ss","14:00:00","-1m","");
+            function.setParameters(params);
+            value = function.execute(result, null);
+            assertTrue(value.equalsIgnoreCase("13:59:00"));
+        }
+        
+        @Test
+        public void testDefault() throws Exception {
+            Collection<CompoundVariable> params = makeParams("","","","");
+            function.setParameters(params);
+            long before = Instant.now().toEpochMilli();
+            value = function.execute(result, null);
+            long after = Instant.now().toEpochMilli();
+            long now = Long.parseLong(value);
+            assertTrue(now >= before && now <= after);
```

Comment:
If we use assertThat with CoreMatchers, we get the values, of the assertions in case of a failing test. It might be nice to use a compare method on Instant instead of converting everything to longs.